### PR TITLE
Refactoring fsm using a stateless library

### PR DESCRIPTION
**Pull request by Anton-Rampage**

### CODE REVIEW

File: pkg/node/state_fsm/sync_state.go [hunk 1]
```diff
@@ -84,77 +109,58 @@ func (a *SyncFsm) Task(task tasks.AsyncTask) (FSM, Async, error) {
...
 	if err := a.baseInfo.peers.UpdateScore(p, score); err != nil {
 		return a, nil, a.Errorf(proto.NewInfoMsg(err))
 	}
-	return noop(a)
+	return a, nil, nil
 }
 
-func (a *SyncFsm) Block(p peer.Peer, block *proto.Block) (FSM, Async, error) {
+func (a *SyncState) Block(p peer.Peer, block *proto.Block) (State, Async, error) {
 	if p != a.conf.peerSyncWith {
```
Comment (nickeskov) [inline pkg/node/state_fsm/sync_state.go:151]: And here

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/node/state_fsm/sync_state.go [hunk 2]
```diff
@@ -32,48 +47,58 @@ func (c conf) Now(tm types.Time) conf {
...
+func (a *SyncState) Transaction(p peer.Peer, t proto.Transaction) (State, Async, error) {
 	return tryBroadcastTransaction(a, a.baseInfo, p, t)
 }
 
-// MicroBlock ignores new microblocks while syncing.
-func (a *SyncFsm) MicroBlock(_ peer.Peer, _ *proto.MicroBlock) (FSM, Async, error) {
-	return a.baseInfo.d.Noop(a)
+func (a *SyncState) PeerError(p peer.Peer, _ error) (State, Async, error) {
+	a.baseInfo.peers.Disconnect(p)
+	if a.conf.peerSyncWith == p {
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/state_fsm/sync_state.go]: Outdated

File: pkg/node/state_fsm/sync_state.go [hunk 3]
```diff
@@ -84,77 +109,58 @@ func (a *SyncFsm) Task(task tasks.AsyncTask) (FSM, Async, error) {
...
-			return NewIdleFsm(a.baseInfo), nil, a.Errorf(err)
-		}
+		return a, nil, a.Errorf(errors.Errorf("unexpected internal task '%d' with data '%+v' received by %s State", task.TaskType, task.Data, a.String()))
 	}
-	return a, nil, nil
 }
 
-func (a *SyncFsm) BlockIDs(peer peer.Peer, signatures []proto.BlockID) (FSM, Async, error) {
+func (a *SyncState) BlockIDs(peer peer.Peer, signatures []proto.BlockID) (State, Async, error) {
 	if a.conf.peerSyncWith != peer {
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/state_fsm/sync_state.go]: Fixed

File: pkg/node/state_fsm/sync_state.go (hunk 1 above)
Comment (alexeykiselev) [reply pkg/node/state_fsm/sync_state.go]: Fixed

File: pkg/node/state_fsm/fsm_common.go [hunk 4]
```diff
@@ -79,28 +129,56 @@ func sendScore(p peer.Peer, storage state.State) {
...
+		return errors.Errorf("[%s] %s", state.String(), e.Error())
 	}
-	internal := sync_internal.InternalFromLastSignatures(extension.NewPeerExtension(p, baseInfo.scheme), lastSignatures)
-	c := conf{
-		peerSyncWith: p,
-		timeout:      30 * time.Second,
+}
+
+func createPermitDynamicCallback(event stateless.Trigger, state *StateData, actionFunc func(...interface{}) (State, Async, error)) stateless.DestinationSelectorFunc {
+	return func(ctx context.Context, args ...interface{}) (stateless.State, error) {
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/state_fsm/fsm_common.go]: Fixed

File: pkg/node/state_fsm/fsm.go [hunk 5]
```diff
@@ -134,14 +144,102 @@ func NewFsm(services services.Services, microblockInterval time.Duration) (FSM,
 		skipMessageList: services.SkipMessageList,
 	}
 
-	b.Scheduler.Reschedule()
+	info.Scheduler.Reschedule()
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/state_fsm/fsm.go]: Fixed

File: pkg/node/state_fsm/fsm_common.go [hunk 6]
```diff
@@ -1,29 +1,83 @@
...
-func newPeer(fsm FSM, p peer.Peer, peers peer_manager.PeerManager) (FSM, Async, error) {
+const (
+	askPeersInterval = 5 * time.Minute
+)
+
+// Set args types for events
+// First arg is Async - return value of event handler
+var (
+	eventsArgsTypes = map[stateless.Trigger][]reflect.Type{
+		PeerErrorEvent:     {reflect.TypeOf(&Async{}), reflect.TypeOf((*peer.Peer)(nil)).Elem(), reflect.TypeOf((*error)(nil)).Elem()},
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/state_fsm/fsm_common.go]: Async will be completely removed in a separate PR.

File: pkg/node/state_fsm/fsm_common.go [hunk 7]
```diff
@@ -1,29 +1,83 @@
...
 
-func newPeer(fsm FSM, p peer.Peer, peers peer_manager.PeerManager) (FSM, Async, error) {
+const (
+	askPeersInterval = 5 * time.Minute
+)
+
+// Set args types for events
+// First arg is Async - return value of event handler
+var (
+	eventsArgsTypes = map[stateless.Trigger][]reflect.Type{
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/state_fsm/fsm_common.go]: Yes, maybe. Added TODO to consider this for further PR.

File: pkg/node/state_fsm/fsm.go [hunk 8]
```diff
@@ -134,14 +144,102 @@ func NewFsm(services services.Services, microblockInterval time.Duration) (FSM,
...
+	return &FSM{
+		fsm:      fsm,
+		baseInfo: info,
+		State:    state,
+	}, t, nil
+}
+
+func (f *FSM) NewPeer(p peer.Peer) (Async, error) {
+	asyncRes := &Async{}
+	err := f.fsm.Fire(NewPeerEvent, asyncRes, p)
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/state_fsm/fsm.go]: Don't get why to pass context? We don't user context for the FSM, it works in background context entirely.

File: pkg/node/node.go [hunk 9]
```diff
@@ -166,27 +166,25 @@ func (a *Node) Run(ctx context.Context, p peer.Parent, internalMessageCh <-chan
 
 	tasksCh := make(chan tasks.AsyncTask, 10)
 
-	fsm, async, err := state_fsm.NewFsm(a.services, a.microblockInterval)
+	fsm, async, err := state_fsm.NewFSM(a.services, a.microblockInterval)
 	if err != nil {
 		zap.S().Errorf("Failed to : %v", err)
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/node.go:152]: Fixed.

File: pkg/node/node.go [hunk 10]
```diff
@@ -196,27 +194,27 @@ func (a *Node) Run(ctx context.Context, p peer.Parent, internalMessageCh <-chan
...
+				async, err = fsm.NewPeer(t.Peer)
 				if err == nil {
-					zap.S().Debugf("[%s] Established connection with %s peer '%s'", fsm, t.Peer.Direction(), t.Peer.ID())
+					zap.S().Debugf("[%s] Established connection with %s peer '%s'", fsm.State.Name, t.Peer.Direction(), t.Peer.ID())
 				}
 			case *peer.InternalErr:
-				fsm, async, err = fsm.PeerError(m.Peer, t.Err)
+				async, err = fsm.PeerError(m.Peer, t.Err)
 			default:
 				zap.S().Warnf("[%s] Unknown info message '%T'", fsm, m)
```
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/node.go]: Already fixed.

File: pkg/node/state_fsm/fsm.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (alexeykiselev) [reply pkg/node/state_fsm/fsm.go]: Added TODO

**Review by alexeykiselev — APPROVED** [review]

**Pull request merged**